### PR TITLE
improve file lookup time

### PR DESCRIPTION
**Pull request by GitMensch**
* don't check any path per library twice
* use canonicalFilePath for the lookup to not stumble over recursive symlinks like the common /usr/bin/X11

for details on the original issue see https://github.com/KDAB/hotspot/pull/579#discussion_r1439844775

### CODE REVIEW

File: app/perfsymboltable.cpp [hunk 1]
```diff
@@ -170,11 +176,16 @@ QFileInfo PerfSymbolTable::findFile(const QString& path, const QString &fileName
                 return fullPath;
         }
     }
+    
+    /* we want to ensure to not check a path twice (overlap of appPath and extraPath,
+       recursive symlinks like /urs/bin/X11, ...) so we store the canonical visited paths
+       for the lookup for each library */
+    QSet<QString> checkedCanonicalPaths;
```
Comment (milianw) [inline app/perfsymboltable.cpp:183]: reserve an arbitrary size here, ideally run it once and sample the checkedCanonicalPaths.capacity() at the end and then just add that here as a magic number. I suspect it's small, so doing something like

```
// reserve an arbitrary small number to remove the first few reallocations and rehashes
checkedCanonicalPaths.reserve(16);
```

**Comment by milianw** [comment]
Please expand the commit message to show:

- what distro you are on, and where the X11 symlink actually points at
- how much time we are saving roughly, before/after

what I don't understand: if it's a recursive symlink, then we'd loop endlessly before, no? how does that not happen I wonder... do you understand that? independently from that, this patch here sounds good to me.

**Review by milianw — CHANGES_REQUESTED** [review]

**Comment by GitMensch** [comment]
> Please expand the commit message to show:
> 
>     * what distro you are on, and where the X11 symlink actually points at

That happens on all Debian/Ubuntu derived ones. Also see https://unix.stackexchange.com/questions/222919/infinitely-nested-directories-within-usr-bin-x11.
How much such I say in the commit message?

>     * how much time we are saving roughly, before/after

For most normal runs (that don't contain duplicate entries or recursive paths) we don't add any time that can be measured easily (with all the file io and other stuff involved). With this I'm down to less than 1 second for this test (a handful of libraries and the recursive symlink in), before I was at > 50 seconds.

Isn't that a bit too chatty to add to the log?

> what I don't understand: if it's a recursive symlink, then we'd loop endlessly before, no? how does that not happen I wonder... do you understand that?

Yes, we use canoncical path, and that "looks" not recursive.

/usr/bin/X11  (X11 pointing to ".") goes to /usr/bin/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11 and then (after a bunch of more) we get into MAX_PATH or something so we don't get any further.

> independently from that, this patch here sounds good to me.

Can you please suggest a commit message?

File: app/perfsymboltable.cpp (hunk 1 above)
Comment (GitMensch) [reply app/perfsymboltable.cpp:183]: will do, 16 is fine

**Comment by milianw** [comment]
> Can you please suggest a commit message?

On Debian/Ubuntu, there is a recursive symlink at `/usr/bin/X11`. Previously we spend a lot
of time traversing this loop until we hit MAX_PATH, which is extremely wasteful. Now we
ensure that we only visit each canonical path once instead. In one test, this makes a difference
between more than 50s to find a handful of libraries down to less than a second with this
patch here.

**Comment by GitMensch** [comment]
Requested changes done.

**Review by milianw — APPROVED** [review]

**Comment by milianw** [comment]
I'm going to bump perfparser in hotspot now too

thanks!

**Pull request merged**